File: backend/app/services/openai_service.py
"""
OpenAI integration for GPT models
Handles LLM calls with token tracking and cost calculation
"""

import json
import time
from typing import Dict, Any, Optional
from openai import AsyncOpenAI, OpenAIError
from sqlalchemy.orm import Session
from app.core.config import settings
from app.models.jd_builder import LLMCall, LLMCallType
import logging

logger = logging.getLogger(__name__)


# Token pricing for OpenAI models (per 1M tokens)
# As of Dec 2024
TOKEN_PRICING = {
    "gpt-4o": {
        "input": 5.00,
        "output": 15.00,
    },
    "gpt-4o-mini": {
        "input": 0.15,
        "output": 0.60,
    },
    "gpt-4-turbo": {
        "input": 10.00,
        "output": 30.00,
    },
}


class OpenAIService:
    """Service for interacting with OpenAI models"""

    def __init__(self):
        self.client = None
        if settings.OPENAI_API_KEY:
            self.client = AsyncOpenAI(api_key=settings.OPENAI_API_KEY)
        self.default_model = "gpt-4o"

    def _calculate_cost(self, model_name: str, input_tokens: int, output_tokens: int) -> Dict[str, float]:
        """Calculate cost based on token usage"""
        # Default to gpt-4o pricing if model not found
        pricing = TOKEN_PRICING.get(model_name, TOKEN_PRICING["gpt-4o"])

        input_cost = (input_tokens / 1_000_000) * pricing["input"]
        output_cost = (output_tokens / 1_000_000) * pricing["output"]
        total_cost = input_cost + output_cost

        return {
            "input_cost": round(input_cost, 6),
            "output_cost": round(output_cost, 6),
            "total_cost": round(total_cost, 6),
        }

    async def invoke_model(
        self,
        prompt: str,
        db: Session,
        user_id: str,
        call_type: LLMCallType,
        system_prompt: Optional[str] = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        model_id: Optional[str] = None,
        job_description_id: Optional[str] = None,
        cv_id: Optional[str] = None,
        cv_parse_detail_id: Optional[str] = None,
        use_toon: bool = True, # Kept for interface compatibility, though TOON is less critical for OpenAI
    ) -> Dict[str, Any]:
        """
        Invoke OpenAI model with token tracking
        """
        if not self.client:
            return {
                "success": False,
                "error": "OpenAI API key not configured",
            }

        start_time = time.time()
        model_name = model_id or self.default_model

        try:
            messages = []
            if system_prompt:
                messages.append({"role": "system", "content": system_prompt})
            
            messages.append({"role": "user", "content": prompt})

            # Invoke OpenAI
            response = await self.client.chat.completions.create(
                model=model_name,
                messages=messages,
                max_tokens=max_tokens,
                temperature=temperature,
            )

            # Extract response
            response_text = response.choices[0].message.content
            
            # Extract token usage
            usage = response.usage
            input_tokens = usage.prompt_tokens
            output_tokens = usage.completion_tokens
            total_tokens = usage.total_tokens

            # Calculate cost
            cost = self._calculate_cost(model_name, input_tokens, output_tokens)

            # Calculate latency
            latency_ms = int((time.time() - start_time) * 1000)

            # Track LLM call in database
            llm_call = LLMCall(
                user_id=user_id,
                job_description_id=job_description_id,
                cv_id=cv_id,
                cv_parse_detail_id=cv_parse_detail_id,
                call_type=call_type,
                model_name=model_name,
                provider="openai",
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                total_tokens=total_tokens,
                input_cost=cost["input_cost"],
                output_cost=cost["output_cost"],
                total_cost=cost["total_cost"],
                prompt_size_chars=len(prompt),
                response_size_chars=len(response_text),
                latency_ms=latency_ms,
                success=True,
            )
            db.add(llm_call)
            db.commit()
            db.refresh(llm_call)

            logger.info(
                f"OpenAI call successful: {call_type.value} | "
                f"Tokens: {input_tokens}/{output_tokens} | "
                f"Cost: ${cost['total_cost']:.6f} | "
                f"Latency: {latency_ms}ms"
            )

            return {
                "success": True,
                "response": response_text,
                "usage": {
                    "input_tokens": input_tokens,
                    "output_tokens": output_tokens,
                    "total_tokens": total_tokens,
                },
                "cost": cost,
                "latency_ms": latency_ms,
                "llm_call_id": str(llm_call.id),
            }

        except OpenAIError as e:
            error_message = str(e)
            logger.error(f"OpenAI API error: {error_message}")

            # Track failed call
            llm_call = LLMCall(
                user_id=user_id,
                job_description_id=job_description_id,
                cv_id=cv_id,
                cv_parse_detail_id=cv_parse_detail_id,
                call_type=call_type,
                model_name=model_name,
                provider="openai",
                input_tokens=0,
                output_tokens=0,
                total_tokens=0,
                input_cost=0.0,
                output_cost=0.0,
                total_cost=0.0,
                prompt_size_chars=len(prompt),
                latency_ms=int((time.time() - start_time) * 1000),
                success=False,
                error_message=error_message,
            )
            db.add(llm_call)
            db.commit()

            return {
                "success": False,
                "error": error_message,
                "response": None,
            }

        except Exception as e:
            error_message = f"Unexpected error: {str(e)}"
            logger.error(error_message)
            
            # Track failed call
            llm_call = LLMCall(
                user_id=user_id,
                job_description_id=job_description_id,
                cv_id=cv_id,
                cv_parse_detail_id=cv_parse_detail_id,
                call_type=call_type,
                model_name=model_name,
                provider="openai",
                input_tokens=0,
                output_tokens=0,
                total_tokens=0,
                input_cost=0.0,
                output_cost=0.0,
                total_cost=0.0,
                prompt_size_chars=len(prompt),
                latency_ms=int((time.time() - start_time) * 1000),
                success=False,
                error_message=error_message,
            )
            db.add(llm_call)
            db.commit()

            return {
                "success": False,
                "error": error_message,
                "response": None,
            }


# Singleton instance
openai_service = OpenAIService()

File: new-backend/src/services/openai.service.ts
import OpenAI from 'openai';
import { env } from '../config/env';
import { LLMCall, LLMCallType } from '../models/llm-call.model';

const TOKEN_PRICING: Record<string, { input: number, output: number }> = {
    "gpt-4o": { input: 5.00, output: 15.00 },
    "gpt-4o-mini": { input: 0.15, output: 0.60 },
    "gpt-4-turbo": { input: 10.00, output: 30.00 },
};

export class OpenAIService {
    private client: OpenAI | null = null;
    private defaultModel = 'gpt-4o-mini';

    constructor() {
        if (env.OPENAI_API_KEY) {
            this.client = new OpenAI({ apiKey: env.OPENAI_API_KEY });
        } else {
            console.warn('OpenAI API Key not found. OpenAIService initialized in MOCK mode.');
        }
    }

    private calculateCost(model: string, inputTokens: number, outputTokens: number) {
        const pricing = TOKEN_PRICING[model] || TOKEN_PRICING['gpt-4o'];
        const inputCost = (inputTokens / 1_000_000) * pricing.input;
        const outputCost = (outputTokens / 1_000_000) * pricing.output;
        return {
            inputCost,
            outputCost,
            totalCost: inputCost + outputCost
        };
    }

    async invokeModel(params: {
        prompt: string;
        userId?: string;
        callType: LLMCallType;
        systemPrompt?: string;
        model?: string;
        jobId?: string; // mapped to job_description_id often
        cvId?: string;
        cvParseDetailId?: string;
    }) {
        if (!this.client) {
            return { success: false, error: 'OpenAI API not configured' };
        }

        const model = params.model || this.defaultModel;
        const startTime = Date.now();
        let usage = { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 };
        let responseText = '';
        let success = false;
        let errorMessage = '';

        try {
            const messages: any[] = [];
            if (params.systemPrompt) messages.push({ role: 'system', content: params.systemPrompt });
            messages.push({ role: 'user', content: params.prompt });

            const completion = await this.client.chat.completions.create({
                model,
                messages,
                temperature: 0.7
            });

            responseText = completion.choices[0].message.content || '';
            usage = completion.usage || usage;
            success = true;

        } catch (err: any) {
            errorMessage = err.message;
            console.error('OpenAI Error:', err);
        }

        const latency = Date.now() - startTime;
        const cost = this.calculateCost(model, usage.prompt_tokens, usage.completion_tokens);

        // Save Log
        if (params.userId) { // Verify we have a user context to save
            try {
                await LLMCall.create({
                    user_id: params.userId,
                    job_description_id: params.jobId,
                    cv_id: params.cvId,
                    cv_parse_detail_id: params.cvParseDetailId,
                    call_type: params.callType,
                    model_name: model,
                    provider: 'openai',
                    input_tokens: usage.prompt_tokens,
                    output_tokens: usage.completion_tokens,
                    total_tokens: usage.total_tokens,
                    input_cost: cost.inputCost,
                    output_cost: cost.outputCost,
                    total_cost: cost.totalCost,
                    prompt_size_chars: params.prompt.length,
                    response_size_chars: responseText.length,
                    latency_ms: latency,
                    success,
                    error_message: errorMessage
                });
            } catch (e) {
                console.error('Failed to log LLM call:', e);
            }
        }

        return {
            success,
            response: responseText,
            usage,
            cost,
            error: errorMessage
        };
    }
}

export const openAIService = new OpenAIService();
